File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.py
from collections import deque

def solution(queue1, queue2):
    total_sum = sum(queue1) + sum(queue2)
    answer_sum = total_sum // 2
    q1_sum = sum(queue1)
    queue1 = deque(queue1)
    queue2 = deque(queue2)
    
    answer = 0
    while queue1 and queue2:
        if q1_sum < answer_sum:
            temp_val = queue2.popleft()
            q1_sum += temp_val
            queue1.append(temp_val)
            answer += 1
        elif q1_sum > answer_sum:
            q1_sum -= queue1.popleft()
            answer += 1
        else:
            return answer

    return -1

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        long totalSum = sum(queue1) + sum(queue2);
        long answerSum = totalSum / 2;
        long q1Sum = sum(queue1);

        Queue<Integer> deque1 = new ArrayDeque<>();
        Queue<Integer> deque2 = new ArrayDeque<>();

        for (int val : queue1) {
            deque1.add(val);
        }

        for (int val : queue2) {
            deque2.add(val);
        }

        int answer = 0;
        while (!deque1.isEmpty() && !deque2.isEmpty()) {
            if (q1Sum < answerSum) {
                int tempVal = deque2.poll();
                q1Sum += tempVal;
                deque1.add(tempVal);
                answer++;
            } else if (q1Sum > answerSum) {
                q1Sum -= deque1.poll();
                answer++;
            } else {
                return answer;
            }
        }

        return -1;
    }

    private long sum(int[] array) {
        long sum = 0L;
        for (int val : array) {
            sum += val;
        }
        return sum;
    }
}
